File: scripts/export.js
#!/usr/bin/env node

import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const PROJECT_ROOT = path.resolve(__dirname, '..');

const DEFAULT_EXCLUDES = new Set([
  '.claude',
  'dist',
  '.git',
  'node_modules',
]);

const BINARY_EXTENSIONS = new Set([
  '.jpg', '.jpeg', '.png', '.gif', '.ico', '.webp',
  '.pdf', '.zip', '.tar', '.gz', '.7z',
  '.woff', '.woff2', '.ttf', '.eot',
  '.exe', '.bin', '.dll', '.so', '.dylib',
]);

/**
 * Checks whether a buffer represents binary content.
 * @param {Buffer} buffer
 * @returns {boolean}
 */
function isBinaryBuffer(buffer) {
  const checkLength = Math.min(buffer.length, 1024);
  for (let i = 0; i < checkLength; i++) {
    if (buffer[i] === 0) {
      return true;
    }
  }
  return false;
}

/**
 * Recursively scans directory and collects relative file paths.
 * @param {string} currentDir
 * @param {string} rootDir
 * @param {Set<string>} excludes
 * @param {string} resolvedOutputFile
 * @returns {string[]}
 */
function collectFiles(currentDir, rootDir, excludes, resolvedOutputFile) {
  const entries = fs.readdirSync(currentDir, { withFileTypes: true });
  /** @type {string[]} */
  const files = [];

  // Sort entries for deterministic output
  entries.sort((a, b) => a.name.localeCompare(b.name, undefined, { numeric: true, sensitivity: 'base' }));

  for (const entry of entries) {
    const fullPath = path.join(currentDir, entry.name);
    const relPath = path.relative(rootDir, fullPath).replace(/\\/g, '/');

    if (excludes.has(entry.name) || excludes.has(relPath)) {
      continue;
    }

    if (path.resolve(fullPath) === resolvedOutputFile) {
      continue;
    }

    if (entry.isDirectory()) {
      files.push(...collectFiles(fullPath, rootDir, excludes, resolvedOutputFile));
    } else if (entry.isFile()) {
      files.push(relPath);
    }
  }

  return files;
}

/**
 * Generates an ASCII tree diagram representing the collected files.
 * @param {string[]} relativeFilePaths
 * @returns {string}
 */
function generateTree(relativeFilePaths) {
  const root = {};

  for (const relPath of relativeFilePaths) {
    const parts = relPath.split('/');
    let current = root;
    for (const part of parts) {
      current[part] = current[part] || {};
      current = current[part];
    }
  }

  /**
   * @param {Record<string, any>} node
   * @param {string} prefix
   * @returns {string[]}
   */
  function formatNode(node, prefix = '') {
    const keys = Object.keys(node).sort((a, b) => {
      const aIsDir = Object.keys(node[a]).length > 0;
      const bIsDir = Object.keys(node[b]).length > 0;
      if (aIsDir !== bIsDir) return aIsDir ? -1 : 1;
      return a.localeCompare(b, undefined, { numeric: true, sensitivity: 'base' });
    });

    const lines = [];
    keys.forEach((key, index) => {
      const isLast = index === keys.length - 1;
      const connector = isLast ? '└── ' : '├── ';
      const childPrefix = isLast ? '    ' : '│   ';
      const isDir = Object.keys(node[key]).length > 0;

      lines.push(`${prefix}${connector}${key}${isDir ? '/' : ''}`);
      if (isDir) {
        lines.push(...formatNode(node[key], `${prefix}${childPrefix}`));
      }
    });
    return lines;
  }

  return formatNode(root).join('\n');
}

/**
 * Exports all project files into a single structured text file.
 * @param {object} options
 * @param {string} [options.output]
 * @param {string[]} [options.excludes]
 */
export function exportProject(options = {}) {
  const outputFileName = options.output || 'project_export.txt';
  const resolvedOutputFile = path.resolve(PROJECT_ROOT, outputFileName);

  const excludes = new Set([...DEFAULT_EXCLUDES, ...(options.excludes || [])]);

  console.log(`Scanning project at: ${PROJECT_ROOT}`);
  console.log(`Excluding directories: ${Array.from(excludes).join(', ')}`);

  const files = collectFiles(PROJECT_ROOT, PROJECT_ROOT, excludes, resolvedOutputFile);

  console.log(`Found ${files.length} included files.`);

  const writeStream = fs.createWriteStream(resolvedOutputFile, { encoding: 'utf8' });

  const separator = '='.repeat(80);
  const subSeparator = '-'.repeat(80);

  // Write header
  writeStream.write(`${separator}\n`);
  writeStream.write(`PROJECT EXPORT: Wireloom\n`);
  writeStream.write(`Generated at: ${new Date().toISOString()}\n`);
  writeStream.write(`Root directory: ${PROJECT_ROOT}\n`);
  writeStream.write(`Excluded patterns: ${Array.from(excludes).join(', ')}\n`);
  writeStream.write(`Total included files: ${files.length}\n`);
  writeStream.write(`${separator}\n\n`);

  // Write directory structure overview
  writeStream.write(`DIRECTORY STRUCTURE OVERVIEW:\n`);
  writeStream.write(`${subSeparator}\n`);
  writeStream.write(generateTree(files));
  writeStream.write(`\n\n`);

  // Write file contents
  writeStream.write(`${separator}\n`);
  writeStream.write(`FILE CONTENTS\n`);
  writeStream.write(`${separator}\n\n`);

  let processedCount = 0;

  for (const relPath of files) {
    const fullPath = path.join(PROJECT_ROOT, relPath);
    const stat = fs.statSync(fullPath);
    const ext = path.extname(relPath).toLowerCase();

    writeStream.write(`${separator}\n`);
    writeStream.write(`FILE: ${relPath} (${stat.size} bytes)\n`);
    writeStream.write(`${separator}\n`);

    if (BINARY_EXTENSIONS.has(ext)) {
      writeStream.write(`[Binary file skipped: ${stat.size} bytes]\n\n`);
    } else {
      const buffer = fs.readFileSync(fullPath);
      if (isBinaryBuffer(buffer)) {
        writeStream.write(`[Binary file detected and skipped: ${stat.size} bytes]\n\n`);
      } else {
        const content = buffer.toString('utf8');
        writeStream.write(content);
        if (!content.endsWith('\n')) {
          writeStream.write('\n');
        }
        writeStream.write('\n');
      }
    }

    processedCount++;
  }

  writeStream.end();

  return new Promise((resolve, reject) => {
    writeStream.on('finish', () => {
      const outStat = fs.statSync(resolvedOutputFile);
      console.log(`Successfully exported ${processedCount} files.`);
      console.log(`Output written to: ${resolvedOutputFile} (${(outStat.size / 1024).toFixed(2)} KB)`);
      resolve(resolvedOutputFile);
    });

    writeStream.on('error', (err) => {
      reject(err);
    });
  });
}

// Direct CLI invocation check
const isCli = process.argv[1] && path.resolve(process.argv[1]) === __filename;
if (isCli) {
  const userOutput = process.argv[2];
  try {
    await exportProject({ output: userOutput });
  } catch (error) {
    console.error('Export failed:', error);
    process.exit(1);
  }
}

File: scripts/export.py
#!/usr/bin/env python3
"""
Wireloom Project Export Script
Exports all project files and folders into a single structured .txt file.
Excludes: .claude, dist, .git, node_modules, and the output file itself.
"""

from datetime import datetime, timezone
from pathlib import Path
import sys

DEFAULT_EXCLUDES = {
    ".claude",
    "dist",
    ".git",
    "node_modules",
}

BINARY_EXTENSIONS = {
    ".jpg", ".jpeg", ".png", ".gif", ".ico", ".webp",
    ".pdf", ".zip", ".tar", ".gz", ".7z",
    ".woff", ".woff2", ".ttf", ".eot",
    ".exe", ".bin", ".dll", ".so", ".dylib",
}


def is_binary_file(file_path: Path) -> bool:
    """Checks if a file is binary by extension or null-byte scanning."""
    if file_path.suffix.lower() in BINARY_EXTENSIONS:
        return True
    try:
        with open(file_path, "rb") as f:
            chunk = f.read(1024)
            return b"\x00" in chunk
    except Exception:
        return True


def collect_files(root_dir: Path, output_file: Path, excludes: set[str]) -> list[Path]:
    """Recursively collects relative paths of all included files."""
    collected: list[Path] = []
    resolved_output = output_file.resolve()

    for path in sorted(root_dir.rglob("*")):
        # Check if any parent or self is in excludes
        rel_parts = path.relative_to(root_dir).parts
        if any(part in excludes for part in rel_parts):
            continue

        if path.resolve() == resolved_output:
            continue

        if path.is_file():
            collected.append(path.relative_to(root_dir))

    return collected


def generate_tree(file_paths: list[Path]) -> str:
    """Generates an ASCII hierarchy tree from a list of relative file paths."""
    tree: dict = {}
    for path in file_paths:
        curr = tree
        for part in path.parts:
            curr = curr.setdefault(part, {})

    def format_node(node: dict, prefix: str = "") -> list[str]:
        lines = []
        keys = sorted(
            node.keys(),
            key=lambda k: (len(node[k]) == 0, k.lower()),
        )
        for i, key in enumerate(keys):
            is_last = i == len(keys) - 1
            connector = "└── " if is_last else "├── "
            child_prefix = "    " if is_last else "│   "
            is_dir = len(node[key]) > 0

            lines.append(f"{prefix}{connector}{key}{'/' if is_dir else ''}")
            if is_dir:
                lines.extend(format_node(node[key], prefix + child_prefix))
        return lines

    return "\n".join(format_node(tree))


def export_project(output_path: str = "project_export.txt") -> None:
    """Main export routine."""
    project_root = Path(__file__).resolve().parent.parent
    out_file = project_root / output_path
    separator = "=" * 80
    sub_separator = "-" * 80

    print(f"Scanning project at: {project_root}")
    print(f"Excluding directories: {', '.join(sorted(DEFAULT_EXCLUDES))}")

    files = collect_files(project_root, out_file, DEFAULT_EXCLUDES)
    print(f"Found {len(files)} included files.")

    with open(out_file, "w", encoding="utf-8", newline="\n") as out:
        # Header
        out.write(f"{separator}\n")
        out.write("PROJECT EXPORT: Wireloom\n")
        out.write(f"Generated at: {datetime.now(timezone.utc).isoformat()}\n")
        out.write(f"Root directory: {project_root}\n")
        out.write(f"Excluded patterns: {', '.join(sorted(DEFAULT_EXCLUDES))}\n")
        out.write(f"Total included files: {len(files)}\n")
        out.write(f"{separator}\n\n")

        # Directory structure overview
        out.write("DIRECTORY STRUCTURE OVERVIEW:\n")
        out.write(f"{sub_separator}\n")
        out.write(generate_tree(files))
        out.write("\n\n")

        # File contents
        out.write(f"{separator}\n")
        out.write("FILE CONTENTS\n")
        out.write(f"{separator}\n\n")

        for rel_path in files:
            full_path = project_root / rel_path
            size = full_path.stat().st_size

            out.write(f"{separator}\n")
            out.write(f"FILE: {rel_path.as_posix()} ({size} bytes)\n")
            out.write(f"{separator}\n")

            if is_binary_file(full_path):
                out.write(f"[Binary file skipped: {size} bytes]\n\n")
            else:
                try:
                    with open(full_path, "r", encoding="utf-8", errors="replace") as f:
                        content = f.read()
                    out.write(content)
                    if not content.endswith("\n"):
                        out.write("\n")
                    out.write("\n")
                except Exception as e:
                    out.write(f"[Error reading file: {e}]\n\n")

    out_size_kb = out_file.stat().st_size / 1024
    print(f"Successfully exported {len(files)} files.")
    print(f"Output written to: {out_file} ({out_size_kb:.2f} KB)")


if __name__ == "__main__":
    target = sys.argv[1] if len(sys.argv) > 1 else "project_export.txt"
    export_project(target)
